File: other/diverta2019-2/D.py
import sys

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


def solve():
    N = int(rl())
    ga, sa, ba = map(int, rl().split())
    gb, sb, bb = map(int, rl().split())
    
    dp1 = [0] * 30000000
    dp1[0] = N
    M = N
    for i in range(N + 1):
        dp1[i + ga] = max(dp1[i + ga], dp1[i] - ga + gb)
        dp1[i + sa] = max(dp1[i + sa], dp1[i] - sa + sb)
        dp1[i + ba] = max(dp1[i + ba], dp1[i] - ba + bb)
        M = max(M, dp1[i])
    
    dp2 = [0] * 30000000
    dp2[0] = M
    ans = M
    for i in range(M + 1):
        dp2[i + gb] = max(dp2[i + gb], dp2[i] - gb + ga)
        dp2[i + sb] = max(dp2[i + sb], dp2[i] - sb + sa)
        dp2[i + bb] = max(dp2[i + bb], dp2[i] - bb + ba)
        ans = max(ans, dp2[i])
    print(ans)


if __name__ == '__main__':
    solve()

File: other/diverta2019-2/D.nim
import sequtils, strutils

proc input(): string =
    return stdin.readLine
proc chmax(num0: var SomeNumber, num1: SomeNumber) =
    num0 = max(num0, num1)

var dp0, dp1: array[30000000, int]

proc solve() =
    let N = input().parseInt
    var ga, sa, ba, gb, sb, bb: int
    (ga, sa, ba) = input().split.map(parseInt)
    (gb, sb, bb) = input().split.map(parseInt)

    dp0[0] = N
    var M = N
    for i in 0..N:
        dp0[i + ga].chmax(dp0[i] - ga + gb)
        dp0[i + sa].chmax(dp0[i] - sa + sb)
        dp0[i + ba].chmax(dp0[i] - ba + bb)
        M.chmax(dp0[i])
    
    dp1[0] = M
    var ans = M
    for i in 0..M:
        dp1[i + gb].chmax(dp1[i] - gb + ga)
        dp1[i + sb].chmax(dp1[i] - sb + sa)
        dp1[i + bb].chmax(dp1[i] - bb + ba)
        ans.chmax(dp1[i])
    echo ans


when is_main_module:
    solve()
